packages/server/src/utils/ldap.ts

```diff
@@ -0,0 +1,74 @@
+import { Client } from 'ldapts';
+import { User } from '../services/database/types';
+import Logger from '@joplin/utils/Logger';
+import { LdapConfig } from './types';
+
+export default async function ldapLogin(email: string, password: string, user: User, config: LdapConfig): Promise<User> {
+	const logger = Logger.create('LDAP');
+
+	const enabled = config.enabled;
+	const userCreation = config.userCreation;
+	const host = config.host;
+	const mailAttribute = config.mailAttribute;
+	const fullNameAttribute = config.fullNameAttribute;
+	const baseDN = config.baseDN;
+	const bindDN = config.bindDN;
+	const bindPW = config.bindPW;
+
+	logger.info(`Starting authentication with Server ${host}`);
+
+	if (password === '') {
+		logger.error('no password entered');
+		throw new Error('no password entered');
+	}
+
+	if (enabled) {
+		let searchResults;
+		const client = new Client({
+			url: host,
+			timeout: 2000,
+			connectTimeout: 1000,
+		});
+
+		if (bindDN.length !== 0) {
+			try {
+				await client.bind(bindDN, bindPW);
+			} catch (ex) {
+				throw new Error('Could not bind to LDAP server.');
+			}
+		}
+
+		try {
+			searchResults = await client.search(baseDN, {
+				filter: `(${mailAttribute}=${email})`,
+				attributes: ['dn', fullNameAttribute],
+			});
+		} catch (ex) {
+			logger.error(`Could not search ldap server ${host}`);
+			return null;
+		}
+
+		if (bindDN.length !== 0) {
+			await client.unbind();
+		}
+
+		try {
+			await client.bind(searchResults.searchEntries[0].dn, password);
+		} catch (ex) {
+			return null;
```

Comment:
Shouldn't silence errors in general. Maybe wrap the error like above and rethrow it?

Reply:
Changed it as suggested. But now I remembered why I did it that way.
Now the user will be confronted with an error like ""
By returning null the user gets "invalid username or password" which is almost always the cause for throwing an error here.

What would you prefer? Maybe returning null but logging the error message?

Reply:
In general we need to distinguish between user errors and server errors. If the call here is to check the user password, is it possible to look at the exception error and return `null` if it's an invalid login, and throw an exception if, for example, the server is unreachable? Does the error object has some code attached to it that we could look at for this?

Reply:
I found the error code for invalid credentials. If an error with this code is thrown, it now will return null and otherwise it will return the error.